WordPress/AbstractThemeSniff.php

```diff
@@ -0,0 +1,1068 @@
+
+	/**
+	 * Array holding the once through WordPress Theme file checks.
+	 *
+	 * @var array
+	 */
+	private static $sniff_helper;
+
+	/**
+	 * Initial setup.
+	 *
+	 * @var array $theme_data contains data from style.css headers.
+	 */
+	public function __construct() {
+		// do a once through pass and get the files.
+		if ( empty( self::$sniff_helper ) || ! is_array( self::$sniff_helper ) || false === self::$sniff_helper ) {
+			$files = $this->get_files();
+			$themedir = $files[0];
+			// Only run this if checking a directory.
+			$themedir_parts = pathinfo( $themedir );
+			if ( isset( $themedir_parts['extension'] ) ) {
+				return;
+			}
+			$themefiles = $this->listdir( $themedir );
+			self::$sniff_helper = $this->once_through( $themedir , $themefiles );
+			// var_dump( self::$sniff_helper );.
+		}
+	}
+
+	/**
+	 * Get the list of files that are being sniffed.
+	 *
+	 * @return array
+	 */
+	private function get_files() {
+		$cli_object = new PHP_CodeSniffer_CLI;
+		$command_line_values = $cli_object->getCommandLineValues();
+		array_multisort( array_map( 'strlen', $command_line_values['files'] ), $command_line_values['files'] );
+		return $command_line_values['files'];
+	}
+
+	// ================== below are functions for the once through
+	/**
+	 * Taken from themecheck
+	 *
+	 * @param string $dir containing directory uri.
+	 */
+	protected function listdir( $dir ) {
+		$files = array();
+		$dir_iterator = new RecursiveDirectoryIterator( $dir );
+		$iterator = new RecursiveIteratorIterator( $dir_iterator, RecursiveIteratorIterator::SELF_FIRST );
+
+		foreach ( $iterator as $file ) {
+			array_push( $files, $file->getPathname() );
+		}
+		return $files;
+	}
+
+	/**
+	 * Taken from themecheck.
+	 * Strip comments from a PHP file in a way that will not change the underlying structure of the file.
+	 *
+	 * @param string $code contains the file contents as a string.
+	 */
+	protected function strip_comments( $code ) {
+		$strip = array( T_COMMENT => true, T_DOC_COMMENT => true );
+		$newlines = array( "\n" => true, "\r" => true );
+		$tokens = token_get_all( $code );
+		reset( $tokens );
+		$return = '';
+		$token = current( $tokens );
+		while ( $token ) {
+			if ( ! is_array( $token ) ) {
+				$return .= $token;
+			} elseif ( ! isset( $strip[ $token[0] ] ) ) {
+				$return .= $token[1];
+			} else {
+				for ( $i = 0, $token_length = strlen( $token[1] ); $i < $token_length; ++$i ) {
+					if ( isset( $newlines[ $token[1][ $i ] ] ) ) {
+						$return .= $token[1][ $i ];
+					}
+				}
+			}
+			$token = next( $tokens );
+		}
+		return $return;
+	}
+
+	/**
+	 * Processes the contents of each theme file.
+	 * Obtain sniff helper data
+	 *
+	 * @param string $themedir directory of theme being checked.
+	 * @param array  $themefiles list of files to be checked.
+	 *
+	 * @return false|array
+	 */
+	public function once_through( $themedir, $themefiles = array() ) {
+		// Initiate the Sniff_helper array.
+		global $sniff_helper;
```

Comment:
I realize that globals are shunned as poor programming but is there a danger in this context to use globals in the way I am using it? In this application it just makes things simpler (for me anyways).
I will look at the direct access approach, to fall in line with a better practice.